cmd/metacache-entries.go

```diff
@@ -58,68 +58,76 @@ func (e metaCacheEntry) hasPrefix(s string) bool {
 	return strings.HasPrefix(e.name, s)
 }
 
-// matches returns if the entries match by comparing their latest version fileinfo.
-func (e *metaCacheEntry) matches(other *metaCacheEntry, bucket string) bool {
+// matches returns if the entries have the same versions.
+// If strict is false we allow signatures to mismatch.
+func (e *metaCacheEntry) matches(other *metaCacheEntry, strict bool) (prefer *metaCacheEntry, matches bool) {
 	if e == nil && other == nil {
-		return true
+		return nil, true
 	}
-	if e == nil || other == nil {
-		return false
+	if e == nil {
+		return other, false
+	}
+	if other == nil {
+		return e, false
 	}
 
-	// This should reject 99%
-	if len(e.metadata) != len(other.metadata) || e.name != other.name {
-		return false
+	// Name should match...
+	if e.name != other.name {
+		if e.name < other.name {
+			return e, false
+		}
+		return other, false
 	}
 
-	eFi, eErr := e.fileInfo(bucket)
-	oFi, oErr := other.fileInfo(bucket)
+	eVers, eErr := e.xlmeta()
+	oVers, oErr := other.xlmeta()
 	if eErr != nil || oErr != nil {
-		return eErr == oErr
+		return nil, false
 	}
 
 	// check both fileInfo's have same number of versions, if not skip
 	// the `other` entry.
-	if eFi.NumVersions != oFi.NumVersions {
-		return false
-	}
-
-	return eFi.ModTime.Equal(oFi.ModTime) && eFi.Size == oFi.Size && eFi.VersionID == oFi.VersionID
-}
-
-// resolveEntries returns if the entries match by comparing their latest version fileinfo.
-func resolveEntries(a, b *metaCacheEntry, bucket string) *metaCacheEntry {
-	if b == nil {
-		return a
-	}
-	if a == nil {
-		return b
-	}
-
-	aFi, err := a.fileInfo(bucket)
-	if err != nil {
-		return b
-	}
-	bFi, err := b.fileInfo(bucket)
-	if err != nil {
-		return a
-	}
-
-	if aFi.NumVersions == bFi.NumVersions {
-		if aFi.ModTime.Equal(bFi.ModTime) {
-			return a
+	if len(eVers.versions) != len(oVers.versions) {
+		eTime := eVers.latestModtime()
+		oTime := oVers.latestModtime()
+		if !eTime.Equal(oTime) {
+			if eTime.After(oTime) {
+				return e, false
+			}
+			return other, false
 		}
-		if aFi.ModTime.After(bFi.ModTime) {
-			return a
+		// Tiebreak on version count.
+		if len(eVers.versions) > len(oVers.versions) {
+			return e, false
 		}
-		return b
+		return other, false
 	}
 
-	if bFi.NumVersions > aFi.NumVersions {
-		return b
+	// Check if each version matches...
+	for i, eVer := range eVers.versions {
+		oVer := oVers.versions[i]
+		if eVer.header != oVer.header {
+			if !strict && eVer.header.matchesNotStrict(oVer.header) {
+				if prefer == nil {
+					if eVer.header.sortsBefore(oVer.header) {
+						prefer = e
+					} else {
+						prefer = other
+					}
+				}
+				continue
+			}
+			if eVer.header.sortsBefore(oVer.header) {
+				return e, false
+			}
+			return other, false
```

Comment:
At this point, shouldn't we return `prefer` if it's non-nil, since `prefer` stores the metacache entry which has a more recent version.
```suggestion
                        if prefer != nil {
                           return prefer, false
                        }
			if eVer.header.sortsBefore(oVer.header) {
				return e, false
			}
			return other, false
```

Reply:
The `prefer` is only mostly interesting if we match non-strict (and only used in this case).

But I don't see anything bad with your suggestion.